src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -715,11 +730,96 @@ async Task Core(MultiConnectSocketAsyncEventArgs internalArgs, Task<IPAddress[]>
                     // Try each address in turn.  We store the last error received, such that if we fail to connect to all addresses,
                     // we can use the last error to represent the entire operation.
                     SocketError lastError = SocketError.NoData;
-                    foreach (IPAddress address in await addressesTask.ConfigureAwait(false))
+                    IPAddress[] addresses = await addressesTask.ConfigureAwait(false);
+                    IPAddress? address = null;
+                    IPAddress? address2 = null;
+                    int nextAddressIndex = 0;
+                    int nextIPv6Addressindex = -1;
+                    AddressFamily currentAddressFamily = AddressFamily.InterNetwork;
+
+                    // We can do parallel connect only if siocket was not specified and when there is at least one address of each AF.
+                    bool parallelConnect = connectAlgorithm == ConnectAlgorithm.Parallel && _currentSocket == null && RemoteEndPoint!.AddressFamily == AddressFamily.Unspecified &&
+                                           nextAddressIndex >= 0 && nextIPv6Addressindex >= 0;
+
+
+                    if (parallelConnect)
+                    {
+                        nextAddressIndex = GetNextAddress(addresses, AddressFamily.InterNetwork, -1);
+                        nextIPv6Addressindex = GetNextAddress(addresses, AddressFamily.InterNetworkV6, -1);
+                        internalArgs.SecondarySaea = new SocketAsyncEventArgs();
+                    }
+
+                    internalArgs.SocketError = SocketError.Success;
+                    while (true)
                     {
+                        if (!parallelConnect)
+                        {
+                            // We simply try addresses in sequence until we either try them all or operation is cancelled.
+                            if (nextAddressIndex >= addresses.Length)
+                            {
+                                break;
+                            }
+                            address = addresses[nextAddressIndex];
+                            nextAddressIndex++;
+                        }
+                        else
+                        {
+                            if (nextAddressIndex == -1 && nextIPv6Addressindex == -1)
```

Comment:
```suggestion
                            if (nextAddressIndex < 0 && nextIPv6Addressindex < 0)
```
The JIT can produce less and more optimized machine-code (if it matters here too, but the change is cheap...)